Add generation types for Create and Update requests

ivanpajon opened the pull request:
#110 This PR adds the generation of types for Create and Update requests.
Also, using the new generated types for create/update I think #65 can be solved due to the proper typing of "file" fields.

(Collections `_authOrigins,_externalAuths,_mfas and _otps` were excluded from this PR as they are internal and I think it is very rare to create/update them via API)

patmood commented:
Hey @ivanpajon! Thanks for the contribution and sorry for the delay. There's a lot of duplication going on with the new output. Could this issue be solved with utility functions? Something like this:

```
type BooksRecord = {
	title: string
}

type Update<T> = Partial<T> & {
    oldPassword?: string
    password?: string
    passwordConfirm?: string
}
type Create<T> = Omit<T, 'id'>

const newBook: Create<BooksRecord> = { title: 'hello' }
const myUpdate: Update<BooksRecord> = { title: 'hello'}
```

ivanpajon commented:
@patmood I have done like this to not introduce breaking changes to the library (and because it was easier than a refactor), but definitely using utility types would be a cleaner approach

ivanpajon commented:
@patmood I updated the PR to implement utility types and reduce duplicated code.

Changes:
- Could be used through Collections enum or Record type
- Utility types now can be used with any collection (including `_authOrigins,_externalAuths,_mfas and _otps`)
- Handle file fields in Record definitions (string | File)
- Any Autodate fields omitted from Create and Update
- Improved TypedPocketBase definition to be more generic and type-safe
- Tests updated

Usage example:
```typescript
export type PostsRecord = {
  id: string
  active: boolean
  creator: string
  created?: IsoAutoDateString
  updated?: IsoAutoDateString
}

const post: CreateBase<PostsRecord> = {
  id: 'UNIQUE_ID',
  active: true,
  creator: 'user',
}
// OR
const post: Create<Collections.Posts> = {
  id: 'UNIQUE_ID',
  active: true,
  creator: 'user',
}

// Same for update with UpdateBase<> and Update<> types
// For Auth collections also can be used CreateAuth<> and UpdateAuth<> types
```

ivanpajon commented:
@patmood I fixed the tests

patmood commented on src/lib.ts:
```diff
@@ -53,15 +57,19 @@ export function generate(
     createCollectionResponses(sortedCollectionNames),
     options.sdk && TYPED_POCKETBASE_COMMENT,
-    options.sdk && createTypedPocketbase(sortedCollectionNames),
+    UTILITY_TYPES,
```
Move this up so it's under `createCollectionResponses`. I think `TypedPocketBase` should be the last thing in the file

patmood commented on src/constants.ts:
```diff
@@ -24,9 +33,68 @@ export type BaseSystemFields<T = never> = {
+\tverified?: boolean
+}`
+
+export const UTILITY_TYPES = `/** Utility types for PocketBase record operations */
```
```suggestion
export const UTILITY_TYPES = `// Utility types for create/update operations
```

patmood commented on src/constants.ts:
```diff
@@ -24,9 +33,68 @@ export type BaseSystemFields<T = never> = {
+export const UTILITY_TYPES = `/** Utility types for PocketBase record operations */
+
+// Helper to determine if a collection is Auth
+type IsAuthCollection<T extends keyof CollectionResponses> =
```
Remove this helper and inline it in `Create` and `Update` types

patmood commented on src/constants.ts:
```diff
@@ -24,9 +33,68 @@ export type BaseSystemFields<T = never> = {
+\tpassword?: string
+\tpasswordConfirm?: string
+\tverified?: boolean
+}`
```
Remove the 4 constants above (BASE_SYSTEM_CREATE_FIELDS_DEFINITION, BASE_SYSTEM_UPDATE_FIELDS_DEFINITION, AUTH_SYSTEM_CREATE_FIELDS_DEFINITION, AUTH_SYSTEM_UPDATE_FIELDS_DEFINITION)

and inline the definitions in `CreateAuth` etc below.

patmood commented on src/utils.ts:
```diff
@@ -35,6 +35,26 @@ export function getSystemFields(type: CollectionRecord["type"]) {
+      return "BaseSystemUpdateFields"
+  }
+}
+
```
Remove these utils. Looks like they are no longer used.

patmood reviewed (commented):
Thanks again for doing this. 

The main theme of my feedback is to simplify. I'd like to keep the output really simple and easy to read and understand after someone generates types. This mostly means removing unnecessary base types and utilities that are only used once or twice and putting them inline.

Could you also please add a section to the readme that shows how you would use the create/update types in app code?

ivanpajon commented:
@patmood I didn't update the readme yet, when I have a moment to write some good examples I will make a commit with them

ivanpajon commented:
@patmood I updated the README

patmood commented:
Thanks there are some merge conflicts, so the branch needs to be updated. Resolve the source file conflicts first, then just run `npm run build` to update the dist file.

Could you also replace `AUTODATE_STRING_TYPE_NAME` with `DATE_STRING_TYPE_NAME` (the unresolved comment).

ivanpajon commented:
I just updated the branch.

Regarding the `AUTODATE_STRING_TYPE_NAME`, I left a comment in the conversation explaining why I can't remove it:

> I needed to do this "trick" in order to be able to exclude any autodate field (without knowing its name in the collection) from the create/update types, as in latest pocketbase versions you can create multiple autodate fields with any name.

patmood commented on test/pocketbase-types-example.ts:
```diff
@@ -222,19 +223,57 @@ export type CollectionResponses = {
+export type CreateBase<T> = {
+	id?: RecordIdString
+} & Omit<{
+	[K in keyof T as T[K] extends IsoAutoDateString | (IsoAutoDateString | undefined) ? never : K]: T[K]
```
Why is this line different for Auth and Base? 
Does it also need to be omitted from the update fields?

Could it be a common helper like this? (not tested)

```
// Autodate fields are not required in updates, because they are set automatically by the server
// File fields are stored as filename strings, but the File object is required when creating/updating.
export type FixTypesForUpdate<T> = {
  [K in keyof T as 
    T[K] extends IsoAutoDateString 
      ? never 
      : K
  ]: T[K] extends FileNameString 
      ? File 
      : T[K] extends FileNameString[] 
        ? File[] 
        : T[K]
}
```

patmood commented on src/fields.ts:
```diff
@@ -27,7 +28,7 @@ export const pbSchemaTypescriptMap = {
   // Dependent on schema
   file: (fieldSchema: FieldSchema) =>
-    fieldSchema.maxSelect && fieldSchema.maxSelect > 1 ? "string[]" : "string",
+    fieldSchema.maxSelect && fieldSchema.maxSelect > 1 ? "string[] | File[]" : "string | File",
```
I think this change should be reverted because it is misleading. The API is always going to store and return a string.

Instead, you could add a helper that adds File types for create/update

patmood reviewed (commented):
Nice work @ivanpajon!

This is looking good and will be a great addition. A few final comments

sean-at-heady commented:
Just adding a comment to say that i would love to see this PR Land. Been a while since last update is there anything that can be done to wrap it up?

patmood commented:
I’m about to start using pocketbase on a new project and will update this
library at the same time

On Thu, 21 Aug 2025 at 12:22 am, sean-at-heady ***@***.***>
wrote:

> *sean-at-heady* left a comment (patmood/pocketbase-typegen#111)
> <https://github.com/patmood/pocketbase-typegen/pull/111#issuecomment-3206636049>
>
> Just adding a comment to say that i would love to see this PR Land. Been a
> while since last update is there anything that can be done to wrap it up?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/patmood/pocketbase-typegen/pull/111#issuecomment-3206636049>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA6YGPGHDWP6IBRJ25BEBQ33OSAEHAVCNFSM6AAAAABUE3Z6OKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTEMBWGYZTMMBUHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

SeanDunford commented:
> I’m about to start using pocketbase on a new project and will update this
> 
> library at the same time
> 
> 
> 
> On Thu, 21 Aug 2025 at 12:22 am, sean-at-heady ***@***.***>
> 
> wrote:
> 
> 
> 
> > *sean-at-heady* left a comment (patmood/pocketbase-typegen#111)
> 
> > <https://github.com/patmood/pocketbase-typegen/pull/111#issuecomment-3206636049>
> 
> >
> 
> > Just adding a comment to say that i would love to see this PR Land. Been a
> 
> > while since last update is there anything that can be done to wrap it up?
> 
> >
> 
> > —
> 
> > Reply to this email directly, view it on GitHub
> 
> > <https://github.com/patmood/pocketbase-typegen/pull/111#issuecomment-3206636049>,
> 
> > or unsubscribe
> 
> > <https://github.com/notifications/unsubscribe-auth/AA6YGPGHDWP6IBRJ25BEBQ33OSAEHAVCNFSM6AAAAABUE3Z6OKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTEMBWGYZTMMBUHE>
> 
> > .
> 
> > You are receiving this because you were mentioned.Message ID:
> 
> > ***@***.***>
> 
> >
> 
> 

I'd be open to collaborating on that.
